File: Aulas/Aula5/Padrao_facade.go
package main

import "fmt"

// Subsistema para fazer café
type Cafeteira struct{}

func (c Cafeteira) MoerGraos() {
    fmt.Println("Moendo grãos de café")
}

func (c Cafeteira) FazerCafe() {
    fmt.Println("Fazendo café")
}

// Subsistema para fazer chá
type Chaleira struct{}

func (c Chaleira) FerverAgua() {
    fmt.Println("Fervendo água")
}

func (c Chaleira) FazerCha() {
    fmt.Println("Fazendo chá")
}

// Fachada
type BebidasFacade struct {
    cafeteira Cafeteira
    chaleira  Chaleira
}

func NewBebidasFacade() BebidasFacade {
    return BebidasFacade{
        cafeteira: Cafeteira{},
        chaleira:  Chaleira{},
    }
}

func (b BebidasFacade) PrepararCafe() {
    fmt.Println("\nPreparando café...")
    b.cafeteira.MoerGraos()
    b.cafeteira.FazerCafe()
}

func (b BebidasFacade) PrepararCha() {
    fmt.Println("\nPreparando chá...")
    b.chaleira.FerverAgua()
    b.chaleira.FazerCha()
}

// Cliente
func main() {
    facade := NewBebidasFacade()
    facade.PrepararCafe()
    facade.PrepararCha()
}

File: Aulas/Aula5/Padrao_facade.java
// Subsistema para fazer café
class Cafeteira {
    public void moerGraos() {
        System.out.println("Moendo grãos de café");
    }

    public void fazerCafe() {
        System.out.println("Fazendo café");
    }
}

// Subsistema para fazer chá
class Chaleira {
    public void ferverAgua() {
        System.out.println("Fervendo água");
    }

    public void fazerCha() {
        System.out.println("Fazendo chá");
    }
}

// Fachada
class BebidasFacade {
    private Cafeteira cafeteira;
    private Chaleira chaleira;

    public BebidasFacade() {
        this.cafeteira = new Cafeteira();
        this.chaleira = new Chaleira();
    }

    public void prepararCafe() {
        System.out.println("\nPreparando café...");
        cafeteira.moerGraos();
        cafeteira.fazerCafe();
    }

    public void prepararCha() {
        System.out.println("\nPreparando chá...");
        chaleira.ferverAgua();
        chaleira.fazerCha();
    }
}

// Cliente
public class Padrao_facade {
    public static void main(String[] args) {
        BebidasFacade facade = new BebidasFacade();
        facade.prepararCafe();
        facade.prepararCha();
    }
}

